### PR TITLE
Migrate function generations to use Syntax Tree instead of modifying AST

### LINKED ISSUE #91: Migrate Azure Functions to use New Compiler API
**Description:**
Currently, Azure functions works by modifying the AST using PackageNode block in compiler extensions. We need to migrate this using codeAnylaze(Project project) block with new compiler apis.

### LINKED ISSUE #111: Disallow using Azure Functions in Submodules
**Description:**
As discussed, it does not make sense to have public faas functions in submodules as they are public and should belong in the default module.

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### LINKED ISSUE #112: Disallow having main function in Azure Functions module
**Description:**
As Azure Functiosn are executed in a different way, having main method does not make sense in azure functions. Having an intermediate azure function solves initializing logic problem. Therefore disable having main function in the default module.

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### PR DESCRIPTION
**Pull request by xlight05**
## Purpose
Earlier, Azure functions works by modifying the AST using PackageNode block in compiler extensions. In this PR we completely avoid editing the AST generating an additional document for generated handler functions instead.

This PR also introduces two levels of additional testing
1. Syntax Tree validation by comparing generated nodes.
2. Runtime validation using mocking http calls of azure functions

Resolves #91

This PR also does few additional validations for the project when it contains Azure Functions
1. Disallow having main function in Azure Functions module  Resolves #112
2. Disallow using Azure Functions in Submodules Resolves #111

## Approach
We iterate through the functions annotated as azure functions and generate intermediate handler functions as to simply the user experience. The generated functions will be written as a separate documented and `Project` will be modified accordingly.

### CODE REVIEW

**Comment by anupama-pathirage** [comment]
@xlight05 The commit messages and PR heading is not according to the best practices. Please correct them.

A properly formed Git commit subject line should always be able to complete the following sentence:

If applied, this commit will **_your subject line here_**

Refer [1] for more details.

[1] https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md#submit-your-contribution

File: azurefunctions-extension-test/src/test/resources/handlers/code/main.bal [hunk 1]
```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (anuruddhal) [inline azurefunctions-extension-test/src/test/resources/handlers/code/main.bal:1]: Fix the header.

File: azurefunctions-extension-test/src/test/java/org/ballerinax/azurefunctions/test/HandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        try {
+            BuildProject project = BuildProject.load(RESOURCE_DIRECTORY.resolve("code"));
+            PackageCompilation compilation = project.currentPackage().getCompilation();
+            DiagnosticResult diagnosticResult = compilation.diagnosticResult();
+            Assert.assertFalse(diagnosticResult.hasErrors());
+            for (ModuleId moduleId : project.currentPackage().moduleIds()) {
+                Module module = project.currentPackage().module(moduleId);
+                for (DocumentId documentId : module.documentIds()) {
+                    Document document = module.document(documentId);
+                    if (document.name().equals("main.bal")) {
```
Comment (anuruddhal) [inline azurefunctions-extension-test/src/test/java/org/ballerinax/azurefunctions/test/HandlerTest.java:51]: Constant should come first.

File: azurefunctions-extension/src/main/java/org/ballerinax/azurefunctions/AzureFunctionVisitor.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        super.visit(modulePartNode);
+    }
+
+    @Override
+    public void visit(ImportDeclarationNode importDeclarationNode) {
+        if (importDeclarationNode.orgName().isEmpty()) {
+            return;
+        }
+        String orgName = importDeclarationNode.orgName().get().orgName().text();
+        if (!orgName.equals("ballerinax")) {
```
Comment (anuruddhal) [inline azurefunctions-extension/src/main/java/org/ballerinax/azurefunctions/AzureFunctionVisitor.java:57]: Constant should come first.

File: azurefunctions-extension/src/main/java/org/ballerinax/azurefunctions/AzureFunctionVisitor.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            return;
+        }
+        if (importDeclarationNode.moduleName().size() != 1) {
+            return;
+        }
+        String moduleName = importDeclarationNode.moduleName().get(0).text();
+        if (moduleName.equals("azure_functions")) {
```
Comment (anuruddhal) [inline azurefunctions-extension/src/main/java/org/ballerinax/azurefunctions/AzureFunctionVisitor.java:64]: Constant should come first.

**Review by anuruddhal — CHANGES_REQUESTED** [review]

**Pull request merged**